### ISSUE TITLE
Cluster Creation in a Region with Local Zones

### ISSUE THREAD

**Issue by vijayansarathy** [weaveworks/eksctl#4734]
Story comment below https://github.com/weaveworks/eksctl/issues/4734#issuecomment-1068961392

---

**What feature/behavior/change do you want?**
When you try to create a cluster in an AWS Region that has one or more local zones enabled, and you are also creating the VPC as part of the cluster creation process, then eksctl passes the subnets in a local zone to the cluster. Unlike regional subnets, EKS can't place network interfaces into Local Zone subnets. As a result, the cluster creation process fails.

**Why do you want this feature?**
AWS customers are trying to use local zones in EKS clusters and are trying to create EKS clusters that include worker nodes in local zones. Currently, they have to create the VPC with regional and local zone subnets as a separate step and then explicitly specify only regional subnets in their configuration file while creating an EKS cluster. They would like to do this with a one step process.

**Comment by nikimanoledaki** [comment]
Hi @vijayansarathy, thanks for flagging this. Could you provide a sample config file (redacted if necessary) and logs so that we can reproduce the error, please?

**Comment by Himangini** [comment]
> Hi @vijayansarathy, thanks for flagging this. Could you provide a sample config file (redacted if necessary) and logs so that we can reproduce the error, please? This would help us come up with a fix for this. Thanks!

@vijayansarathy Are you able to provide a sample config file?

**Comment by 4BitBen** [comment]
@Himangini @nikimanoledaki here is a sample file. I enabled the local zones in us-west on my account.

```
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: test-local-zones
  region: us-west-2
  version: "1.21"

availabilityZones: ["us-west-2b", "us-west-2c", "us-west-2-den-1a", "us-west-2-sea-1a"]

iam:
  withOIDC: true
  serviceAccounts:
    - metadata:
        name: cluster-autoscaler
        namespace: kube-system
      wellKnownPolicies:
        autoScaler: true

managedNodeGroups:
- name: nodegroup
  desiredCapacity: 1
  minSize: 1
  maxSize: 3
  instanceType: m5.xlarge
  availabilityZones: ["us-west-2b", "us-west-2c"]
nodeGroups:
- name: nodegroup-den
  desiredCapacity: 1
  minSize: 1
  maxSize: 3
  instanceType: c5d.2xlarge
  availabilityZones: ["us-west-2-den-1a"]
- name: nodegroup-sea
  desiredCapacity: 1
  minSize: 1
  maxSize: 3
  instanceType: c5d.2xlarge
  availabilityZones: ["us-west-2-sea-1a"]
```

here is a sample run (with some of it redacted):

```
$ eksctl create cluster -f eksctl-us-west-2-local-zones.yaml
2022-02-14 07:55:50 [ℹ]  eksctl version 0.77.0
...
2022-02-14 07:55:51 [ℹ]  using Kubernetes version 1.21
2022-02-14 07:55:51 [ℹ]  creating EKS cluster "test-local-zones" in "us-west-2" region with managed nodes and un-managed nodes
2022-02-14 07:55:51 [ℹ]  3 nodegroups (nodegroup, nodegroup-den, nodegroup-sea) were included (based on the include/exclude rules)
2022-02-14 07:55:51 [ℹ]  will create a CloudFormation stack for cluster itself and 2 nodegroup stack(s)
2022-02-14 07:55:51 [ℹ]  will create a CloudFormation stack for cluster itself and 1 managed nodegroup stack(s)
2022-02-14 07:55:51 [ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-west-2 --cluster=test-local-zones'
...
2022-02-14 07:55:51 [ℹ]  building cluster stack "eksctl-test-local-zones-cluster"
2022-02-14 07:55:52 [ℹ]  deploying stack "eksctl-test-local-zones-cluster"
2022-02-14 07:56:22 [ℹ]  waiting for CloudFormation stack "eksctl-test-local-zones-cluster"
2022-02-14 07:56:52 [ℹ]  waiting for CloudFormation stack "eksctl-test-local-zones-cluster"
2022-02-14 07:56:53 [✖]  unexpected status "ROLLBACK_IN_PROGRESS" while waiting for CloudFormation stack "eksctl-test-local-zones-cluster"
2022-02-14 07:56:53 [ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
2022-02-14 07:56:54 [✖]  AWS::EC2::Route/PublicSubnetRoute: CREATE_FAILED – "Resource creation cancelled"
2022-02-14 07:56:54 [✖]  AWS::EC2::NatGateway/NATGateway: CREATE_FAILED – "Resource creation cancelled"
2022-02-14 07:56:54 [✖]  AWS::EKS::Cluster/ControlPlane: CREATE_FAILED – "Resource handler returned message: \"Cannot create cluster 'test-local-zones' because us-west-2-sea-1a, the targeted availability zone, does not currently have sufficient capacity to support the cluster. Retry and choose from these availability zones: us-west-2a, us-west-2b, us-west-2c, us-west-2d (Service: Eks, ...)"
2022-02-14 07:56:54 [!]  1 error(s) occurred and cluster hasn't been created properly, you may wish to check CloudFormation console
2022-02-14 07:56:54 [ℹ]  to cleanup resources, run 'eksctl delete cluster --region=us-west-2 --name=test-local-zones'
2022-02-14 07:56:54 [✖]  ResourceNotReady: failed waiting for successful resource state
```

I would always get a `the targeted availibility zone does not currently have sufficient capacity to support the cluster` regardless of which local zone I tried. So, I think the sufficient capacity error message might not be accurate. Hope this is enough to reproduce on your end.

**Comment by Himangini** [comment]
@aclevername @nikimanoledaki will fill out more information.

**Comment by aclevername** [comment]
summarising the above into a story

### Description
AWS support a set of zones called "local zones" that are different to traditional availability zones. For example:
- Traditional: `us-west-2a`
- Local: `us-west-2-lax-1`

EKS is not supported in local zones, meaning you cannot create clusters or managed nodegroups in the zones. However you can created unmanaged nodegroups in the zones. 

Today if you want to use local zones with eksctl you have to create the VPC yourself and then provide that to eksctl when creating the nodegroups, but not the cluster. This is because if you provide any local zones to `eksctl create cluster` we will attempt to create the cluster in the zone, which throws an error.

This story is to add support so that users can provide a set of local zones, in addition to the required (min 2) regular availability zones, so that eksctl will create the VPC with them, but not the cluster/managed nodegroups.


### Scope
- Add support for `eksctl create cluster` where local zones are provided
  - The VPC should be created with the zones
  - Cluster/managed nodegroups should not

### User story


As a User I want to be able to specify additional local zones when creating a cluster so that I can provision unmanaged nodegroups in those zones.

### Acceptance Criteria


Given a cluster config that specifies a set of non-local zone, local zones and an unmanaged nodegroup
Then it should create a VPC with the all the zones included, and the cluster in the non-local zones
And create the unmanaged nodegroups in the local and non-local zones

### Technical Notes
We can either add support for specifying local zones in the traditional `availabilityZones` section, e.g.:
```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: cluster-23

availabilityZones:
- us-west-2a
- us-west-2b
- us-west-2-lax-1
```

Or add a new field:
```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: cluster-23

availabilityZones:
- us-west-2a
- us-west-2b

localZones:
- us-west-2-lax-1
```


We just have to make sure that we enforce the minimum amount of non-local zones that we required today, which is 2/3 depending on the region. `DescribeAvailabilityZones` [API call](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-availability-zones.html) will return what type a zone is, which is useful:
```
ZoneType -> (string)

The type of zone. The valid values are availability-zone , local-zone , and wavelength-zone .

```

**Comment by Skarlso** [comment]
For the record, I like this one better:

```
localZones:
- us-west-2-lax-1
```

**Comment by Himangini** [comment]
We'll add a separate field for local zones

**Comment by nikimanoledaki** [comment]
One important thing that was missed was that subnets currently have a very tightly coupling with AZs. In fact, we have a 1:1 mapping between AZs and subnets throughout the codebase. For example, see [here](https://github.com/weaveworks/eksctl/blob/main/pkg/apis/eksctl.io/v1alpha5/vpc.go#L30:L49) and [here](https://github.com/weaveworks/eksctl/blob/main/pkg/apis/eksctl.io/v1alpha5/vpc.go#L274-L328). There are more instances like these where AZ is the default way of setting the zone that the subnet should be created in. 

The alternative would be to add a new field such as `LocalZone`. However, we default to setting AZ as the default zone type for creating subnets throughout the codebase, so allowing for extra fields would mean that this would add a significant amount of logic every time that a new type of zone is added where Subnets can be created in. 

The most straightforward way to change this is to allow for other types of zones is to rename the AZ field to Zone. However, this is a user-facing field, so while we can already add a new field Zone, we also need to deprecate the field AZ. Hence the following ticket: https://github.com/weaveworks/eksctl/issues/5087

**Comment by nikimanoledaki** [comment]
Ah yep the key and `az` field should be the same, you're right :) 

Huh. You raise a very interesting point about the key of the map. Yes, that key is the AZ/Zone. Would you say that could be `Option 3` - remove the dependency on the `AZ` field and instead read/set the key as the AZ/Zone? 

We would still need to deprecate the `AZ` field. That would also require quite a lot of refactoring as we would have 2 separate streams of logic - one where we read/set the `AZ` field by default, and one where we read/set the key to find out the zone.

**Comment by aclevername** [comment]
> Ah yep the key and `az` field should be the same, you're right :)

:+1: 

> Thanks @aclevername. :) You raise a very interesting point about the key of the map. Yes, that key is the AZ/Zone. Would you say that could be `Option 3` - remove the dependency on the `AZ` field and instead read/set the key as the AZ/Zone?

Yeah its defo an option :grin: its almost 2x breaking changes though (removing `az` field and mandating the key be the zone), but it almost feels like an even cleaner end outcome since it feels weird that we support setting the AZ via the key or the `az` field. I wonder if theirs any other reason why we support these two different methods

> We would still need to deprecate the `AZ` field. That would also require quite a lot of refactoring as we would have 2 separate streams of logic until we deprecate `AZ` - one where we read/set the `AZ` field by default, and one where we read/set the key to find out the zone.
> 
> Correct! We can have `AZ` with the message that it will be deprecated, add `Zone` at the same time, and validate that not both are used.

:ok_hand: makes sense. Seems good to me

**Comment by nikimanoledaki** [comment]
Ah no! The key of the map does not necessarily have to be an AZ. It can be named something else such as `public-1` - either way, it becomes the name of the subnet. So it can sometimes correspond to the AZ of the subnet, but it's not guaranteed to be the AZ name.

**Comment by aclevername** [comment]
Chatting with @nikimanoledaki on slack we realised option 3 wouldn't work, as the key name can be used for referencing the subnets in nodegroups. E.g.:
```
vpc:
  id: vpc-0c24b1cc77da8f463
  cidr: "192.168.0.0/16"
  subnets:
    public:
      public-one: 
        id: subnet-0324095a4f2188967
        # az: us-west-2a # any unambiguous combination of these is allowed
        # cidr: 192.168.64.0/19
      public-two: 
        id: subnet-083814b087bad140b
nodeGroups:
  - name: ng-1
    instanceType: m5.xlarge
    desiredCapacity: 2
    subnets:
      - public-one
      - public-two
```

**Comment by cPu1** [comment]
I agree with adding a new top-level field `localZones`, instead of adding support for local zones in `availabilityZones`. 


As for supporting a local zone at the subnet level, I'd suggest adding a new field `subnet.localZone`, instead of deprecating and renaming `subnet.AZ` to `subnet.Zone` for these reasons:

a. I believe only a few users will be using local zones and that does not warrant deprecating and renaming a widely-used feature (`subnet.az`).

b. AWS also supports Wavelength Zones that are different from Availability Zones and Local Zones. Having a `Zone` field may give users the impression that it also takes Wavelength Zones in addition to Local and Availabilty Zones. (Although Wavelength Zones are configured differently from Local Zones).

c. Since we've reached consensus on adding a new top-level field `localZones` (instead of overloading `availabilityZones`), `subnet.localZone` is more consistent than `subnet.zone`.

**Comment by Skarlso** [comment]
Optionally, how about this. We provide the new top-level LocalZone setting. We use that to create the VPC and the subnets.
Then, in the nodegroup section, we continue to use the given availability zones that can contain local zones. But, we parse those out because LocalZones look differently from availability zones because of the third `-`. So the usage remains clear and people can use the same field to specify them. Internally, we will, of course, only use local zones for unmanaged nodegroups.

**Comment by Skarlso** [comment]
Oh, and as a heads up, we'll have to coordinate with AWS to change the docs on this part once this is done:

>Nodes

>You can't create managed node groups in AWS Local Zones with Amazon EKS. You must create self-managed nodes using the Amazon EC2 API or AWS CloudFormation. Do not use eksctl to create your cluster or nodes in Local Zones. For more information, see [Self-managed nodes](https://docs.aws.amazon.com/eks/latest/userguide/worker.html).

**Comment by Skarlso** [comment]
After re-reading this a couple of times and the added docs and the EKS networking guide, I agree with Chetan but with a slight modification. @cPu1 what do you think about adding the subnet.localZone setting only for the unmanaged nodegroup section of the configuration? This way, we can be sure that managed nodegroup will not use those local zones when be created.

**Comment by cPu1** [comment]
> @cPu1 what do you think about adding the subnet.localZone setting only for the unmanaged nodegroup section of the configuration?

Sounds right to me, as this will also make the generated schema more accurate. (I wasn't suggesting adding `localZone` to the shared subnet struct).

**Comment by Skarlso** [comment]
>(I wasn't suggesting adding localZone to the shared subnet struct).

Ah, my bad, gotcha.

**Comment by Skarlso** [comment]
After some investigation of the code and how this thing works, I have one observation to make here.

This will not work with CloudFormation during `cluster create`. The CloudFormation template **must not** contain the subnet which contains the local-zone mapping. Otherwise, it will fail creation because of the error above. There is no way to flag the subnet as **do not use** or anything like that. This means the creation must be a post creation extension of the VPC. But! The calculation of the CIDR and the number of subnets still **must** consider this AZ, too; otherwise, the overall numbers will be incorrect, ie. the CIDR prefix and the subnet divide count.

This is a hard nut to crack.

**Comment by cPu1** [comment]
I spent some time looking into this issue. While the implementation is feasible, the scope of the issue needs to be refined. 

Addressing this point from the scope:

>Unmanaged nodegroups can be created in all of the zones

I believe when a user wishes to create a nodegroup in local zones, they'd want all nodes to be in local zones, and not in a mix of both local and availability zones. Therefore, the decision to create a nodegroup in local zones should be explicit, by specifying `nodeGroup.localZones`. To create a nodegroup, a user would specify the desired local zones in `nodeGroup.localZones` and eksctl would pick subnets from those zones. `nodeGroup.localZones` should be a subset of `ClusterConfig.localZones` otherwise eksctl will throw an error. If the subnets in local zones were created by eksctl
and no subnets, AZs or local zones are specified in the nodegroup config, eksctl will not use a local zone.


Regarding https://github.com/weaveworks/eksctl/issues/4734#issuecomment-1091542339, which talks about modifying the VPC config to accommodate local zones, eksctl should not allow users to specify pre-existing subnets in local zones in `vpc.subnets` because local zones can't be treated the same as availability zones as they can't be passed to the control plane (I also ran into capacity issues when creating a NAT gateway in a local zone). 

Lastly, I believe we should let users supply a pre-existing subnet in a local zone when creating a nodegroup. This will not be supported via the `localZones` field but by supplying the subnet ID(s) in `nodeGroup.subnets`. If `nodeGroup.subnets` contains a mix of subnets in both local and availability zones, I think we should allow that but with a warning message. eksctl will also validate that `managedNodeGroup.subnets` contains only subnets in AZs and not local or other types of zones.
